internal/cmd/cli/cleanup/cleanup.go

```diff
@@ -123,13 +124,97 @@ func ClusterRegistrations(ctx context.Context, client Getter, opts Options) erro
 
 		clusterKey := types.NamespacedName{Namespace: cr.Namespace, Name: cr.Status.ClusterName}
 		if _, found := seen[clusterKey]; !found {
-			logrus.Infof("Deleting granted cluster registration without cluster %s/%s", cr.Namespace, cr.Name)
-			if err := clusterRegistration.Delete(cr.Namespace, cr.Name, nil); err != nil && !apierrors.IsNotFound(err) {
-				logrus.Errorf("Failed to delete cluster registration %s/%s: %v", cr.Namespace, cr.Name, err)
+			logger := logger.WithValues("namespace", cr.Namespace, "name", cr.Name)
+			logger.Info("Deleting granted cluster registration without cluster")
+			if err := cl.Delete(ctx, &cr); err != nil && !apierrors.IsNotFound(err) {
+				logger.Error(err, "Failed to delete cluster registration")
 			}
 			continue
 		}
 	}
 
 	return nil
 }
+
+func GitJobs(ctx context.Context, cl client.Client, bs int) error {
+	logger := log.FromContext(ctx).WithName("cleanup-jobs").WithValues("batchSize", bs)
+
+	list := &batchv1.JobList{}
+	if err := cl.List(ctx, list, client.Limit(bs)); err != nil {
+		return err
+	}
+	logger.Info("Listing resources", "jobs", len(list.Items))
+
+	jobs := list.Items
+
+	for list.Continue != "" {
+		if err := cl.List(ctx, list, client.Limit(bs), client.Continue(list.Continue)); err != nil {
+			return err
+		}
+		logger.Info("Listing more resources", "jobs", len(list.Items))
+
+		jobs = append(jobs, list.Items...)
+	}
+
+	if err := cleanupGitJobs(ctx, logger, cl, jobs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func cleanupGitJobs(ctx context.Context, logger logr.Logger, cl client.Client, jobs []batchv1.Job) error {
+	// jobs by namespace, gitrepo
+	gitjobs := map[string]map[string][]batchv1.Job{}
+	// gitrepos with running jobs
+	running := map[string]map[string]struct{}{}
+
+	for _, job := range jobs {
+		if job.OwnerReferences == nil {
+			continue
+		}
+		for _, or := range job.OwnerReferences {
+			if or.Kind == "GitRepo" && or.APIVersion == "fleet.cattle.io/v1alpha1" {
+				if job.Status.Succeeded != 1 || job.Status.CompletionTime == nil {
+					if running[job.Namespace] == nil {
+						running[job.Namespace] = map[string]struct{}{}
+					}
+					running[job.Namespace][or.Name] = struct{}{}
+				} else {
+					if gitjobs[job.Namespace] == nil {
+						gitjobs[job.Namespace] = map[string][]batchv1.Job{}
+					}
+					gitjobs[job.Namespace][or.Name] = append(gitjobs[job.Namespace][or.Name], job)
+				}
+				break
+			}
+		}
+	}
+
+	for ns, gitrepos := range gitjobs {
+		for gitrepo, jobs := range gitrepos {
+			sort.Slice(jobs, func(i, j int) bool {
+				return jobs[j].Status.CompletionTime.Before(jobs[i].Status.CompletionTime)
+			})
+
+			// if there is a running job delete all the jobs in the
+			// list, otherwise all but the newest
+			start := 1
+			if _, ok := running[ns][gitrepo]; ok {
+				start = 0
+			}
+
+			logger.V(1).Info("Deleting jobs for gitrepo", "n", len(jobs)-start, "namespace", ns, "gitrepo", gitrepo)
```

Comment:
Would we want to make this more precise, or would we simply rely on logs from requests made by the caller function?
```suggestion
			logger.V(1).Info("Deleting jobs for gitrepo", "n", len(jobs)-start, "total", len(jobs), "namespace", ns, "gitrepo", gitrepo)
```

Reply:
I'm not sure what you want to log?

Reply:
The idea was to log the total amount of jobs being processed, to be able to follow the deletion job's progress through logs, but that's not critical I guess.